### PR TITLE
[Filebeat] kafka v2 using parsers

### PR DESCRIPTION
**Pull request by mjmbischoff**
## What does this PR do?

Moving it forward to version 2 of the input code removes tech dept. Adding parsers allows the use of parsers, allowing the payload to be interpenetrated in a flexible way which for example fixes the issue outlined in https://github.com/elastic/beats/pull/26833 See also https://github.com/elastic/beats/issues/26130 & https://github.com/elastic/beats/issues/15324

## Why is it important?

Addresses tech dept, allows the kafka input to be used in more setups.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

See the integration tests for the kafka input

## Related issues

- Relates https://github.com/elastic/beats/issues/26130 https://github.com/elastic/beats/issues/15324
- Superseds https://github.com/elastic/beats/pull/26833

## Use cases

Having parsers allows the payload to be interpreted in a more flexible way, for example json can be picked up as structured data. This means that if the data is in the correct structure it can be picked up by a module.  https://github.com/elastic/beats/pull/27154 and https://github.com/elastic/beats/pull/26862 would also allows us to further correct things on the FB side but this PR allows preprocessing to happen in an external logstash or filebeat in front of the filebeat with the module

### CODE REVIEW

**Review by faec — COMMENTED** [review]
This looks really good! The only obvious thing I see that needs work (other than updating generated files to make the linter happy) is handling of `expandEventListFromField`. Otherwise everything looks like a clear improvement over the previous version :-)

**Comment by mjmbischoff** [comment]
Think should be pretty good now, I do still need to add documentation and update the description of the PR

I adapted the Ack for `expandEventListFromField` to only ack when all messages are send to elasticsearch. This to avoid it not sending at least once if messages spawn across two batches and filebeat fails. Also we seem to commit the offset multiple times which requires us to obtain ownership of a mutex, whcih feels a little expensive also as a coincidence https://twitter.com/gwenshap/status/1429147326455508992 showed up on my feed which gives me an idea for another PR.

Picked a reader over a parser because if we introduce generic behavior we might want different name / location `expandEventListFromField` config.

**Comment by mjmbischoff** [comment]
couldn't reuse `input-filestream-reader-options.asciidoc` as it wording is a little log specific

File: filebeat/input/kafka/config.go [hunk 1]
```diff
@@ -215,7 +216,6 @@ func newSaramaConfig(config kafkaInputConfig) (*sarama.Config, error) {
 	)
 
 	if err := k.Validate(); err != nil {
-		logp.Err("Invalid kafka configuration: %v", err)
```
Comment (kvch) [inline filebeat/input/kafka/config.go:218]: :tada: :tada: :tada:

File: filebeat/input/kafka/input.go [hunk 2]
```diff
@@ -21,96 +21,161 @@ import (
...
-	connector channel.Connector,
-	inputContext input.Context,
-) (input.Input, error) {
+func (input *kafkaInput) Name() string { return pluginName }
 
-	config := defaultConfig()
-	if err := cfg.Unpack(&config); err != nil {
-		return nil, errors.Wrap(err, "reading kafka input config")
-	}
+func (input *kafkaInput) Test(_ input.TestContext) error {
```
Comment (kvch) [inline filebeat/input/kafka/input.go:85]: Is there a way maybe we could test connecting to Kafka?

File: filebeat/input/kafka/input.go [hunk 3]
```diff
@@ -325,33 +283,162 @@ func (h *groupHandler) ack(message *sarama.ConsumerMessage) {
...
+		},
+		Meta: common.MapStr{
+			"ackHandler": ackHandler,
+		},
+	}
+}
+
 // parseMultipleMessages will try to split the message into multiple ones based on the group field provided by the configuration
-func (h *groupHandler) parseMultipleMessages(bMessage []byte) []string {
+func parseMultipleMessages(bMessage []byte, field string, log *logp.Logger) []string {
```
Comment (kvch) [inline filebeat/input/kafka/input.go:428]: maybe this can be a function of `listFromFieldReader`?

File: filebeat/input/kafka/input.go (hunk 2 above)
Comment (mjmbischoff) [reply filebeat/input/kafka/input.go]: Added a check that retrieves all topics and checks it against the configured topics.

File: filebeat/input/kafka/config.go (hunk 1 above)
Comment (mjmbischoff) [reply filebeat/input/kafka/config.go:218]: This is good I think / hope? :sweat_smile:

File: filebeat/input/kafka/input.go (hunk 3 above)
Comment (mjmbischoff) [reply filebeat/input/kafka/input.go]: ah yes I initially  left it unchanged from the original code but this is better, done.

File: filebeat/input/kafka/config.go (hunk 1 above)
Comment (kvch) [reply filebeat/input/kafka/config.go:218]: Yes :D

File: filebeat/input/kafka/kafka_integration_test.go [hunk 4]
```diff
@@ -357,3 +451,27 @@ func writeToKafkaTopic(
 		t.Fatal(err)
 	}
 }
+
+func run(t *testing.T, cfg *common.Config, client *beattest.ChanClient) (func(), *kafkaInput) {
```
Comment (kvch) [inline filebeat/input/kafka/kafka_integration_test.go:455]: Please change the order of the returned value. The cancel function is usually the second value.

File: filebeat/input/kafka/input_test.go [hunk 5]
```diff
@@ -21,16 +21,13 @@ package kafka
...
 )
 
 func TestNewInputDone(t *testing.T) {
-	config := common.MapStr{
-		"hosts":    "localhost:9092",
-		"topics":   "messages",
-		"group_id": "filebeat",
-	}
-	inputtest.AssertNotStartedInputCanBeDone(t, NewInput, &config)
+	//config := common.MapStr{
```
Comment (kvch) [inline filebeat/input/kafka/input_test.go:27]: What's up with this?

**Review by kvch — COMMENTED** [review]
Minor comments, we are going in the right direction. Thanks!

File: filebeat/input/kafka/input_test.go (hunk 5 above)
Comment (mjmbischoff) [reply filebeat/input/kafka/input_test.go:27]: ah yes, wanted to discuss this. For the old style input there was the inputtest.AssertNotStartedInputCanBeDone(t, NewInput, &config) call I was wondering if under v2 input this test still makes sense and if there is an input.v2 equivalent (couldn't find it).

File: filebeat/input/kafka/kafka_integration_test.go (hunk 4 above)
Comment (mjmbischoff) [reply filebeat/input/kafka/kafka_integration_test.go]: Fixed in https://github.com/elastic/beats/pull/27335/commits/ad521cd749fef1e5cfac8ac75eca3325d3e5cb64

**Review by faec — APPROVED** [review]
This looks great, thank you!

**Comment by mjmbischoff** [comment]
Thanks, @faec & @kvch !

**Pull request merged**